modules/common/react/lib/utils/components.ts

```diff
@@ -0,0 +1,174 @@
+import React from 'react';
+
+/**
+ * Adds the `as` to the style interface to support `as` in styled components
+ * This type and usage should be removed with Emotion 11
+ */
+export type StyledType = {
+  as?: React.ElementType;
+};
+
+/**
+ * Generic component props with "as" prop
+ * @template P Additional props
+ * @template T React component or string element
+ */
+export type PropsWithAs<P, T extends React.ElementType> = P &
+  Omit<React.ComponentProps<T>, 'as' | 'state' | keyof P>;
+
+/**
+ * Component type that allows for `as` to change the element or component type.
+ * Passing `as` will correctly change the allowed interface of the JSX element
+ */
+export type Component<T extends React.ElementType, P> = {
+  <TT extends React.ElementType>(props: PropsWithAs<P, TT> & {as: TT}): JSX.Element;
```

Comment:
The double T here (`TT`) is new to me. What does this do??

Reply:
It's just another generic, since T is already being used elsewhere.

Reply:
It is used because it is the same intended type as `T`, but is a different type in this instance. `Component` is an overloaded type:

```tsx
const Foo = Component<'div', Props>

<Foo /> // hits the second overload which means it has all the inherit props of HTMLDivElement
<Foo as="button" /> // matches the first overload and means it has all the inherit props of HTMLButtonElement
```

Reply:
All that's saying is "whatever component is provided will extend React.ElementType, use its own props + `as`, where `as` is the type of component being provided."

Reply:
The double `TT` is another generic.